Produce consume tutorial in scala (#459)
confluentinc/kafka-tutorials: _includes/tutorials/produce-consume-lang/scala/markup/dev/6_0-consume-narrative.adoc

The following Scala class will be the entry point of the consumer application.
At `src/main/scala/io/confluent/developer/Consumer.scala` add:

+++++
<pre class="snippet"><code class="java">{%
    include_raw tutorials/produce-consume-lang/scala/code/src/main/scala/io/confluent/developer/Consumer.scala
%}</code></pre>
+++++

Lets’s describe the keys pieces of this program.

Reminder: we use an externalised config and this app loads its config from the _consumer_ bloc of the `application.conf`

A `Consumer#conusme` function covers most of the record consumption.

- `Consumer#conusme` takes a
`https://kafka.apache.org/25/javadoc/org/apache/kafka/clients/consumer/KafkaConsumer.html[KafkaConsumer]`
and gives back a collection of `Book`.

- `Consumer#consume` call the `KafkaConsumer#poll` and gets a `ConsumerRecords` filed with as much books as the
`fetch.max.bytes` Consumer Config allows.

- `Consumer#consume` extracts the value from each records. The consumed books will filled `bookMap`, our book collection
of type `mutable.Map[String, Book]`.

> - We also have two functions `getCount` and `getBooks` just to define two HTTP routes.

Reviewer: Same thing here put the `Consumer.consume` in a code block with callouts in the narrative.

> The following Scala class will be the entry point of the consumer application.
> At `src/main/scala/io/confluent/developer/Consumer.scala` add:

Reviewer 2: ```suggestion
The following Scala class defines our Consumer application.

Add the following class to the file `src/main/scala/io/confluent/developer/Consumer.scala`
```

+++++
<pre class="snippet"><code class="java">{%
    include_raw tutorials/produce-consume-lang/scala/code/src/main/scala/io/confluent/developer/Consumer.scala
%}</code></pre>
+++++

Lets’s describe the keys pieces of this program.

> NOTE: We use an externalised config and this app loads its config from the _consumer_ bloc of the `application.conf`

Reviewer 2: ```suggestion
NOTE: The Consumer application loads its configuration from the _consumer_ bloc of the `application.conf`
```

A `Consumer#conusme` function covers most of the record consumption.

[source,scala]
----
def consume(consumer: KafkaConsumer[Bytes, Book]): Vector[Book] = { //<1>

  val books: ConsumerRecords[Bytes, Book] = consumer.poll((1 second) toJava) //<2>

  books.asScala.toVector.map(_.value()) //<3>
}
----

<1> `Consumer#conusme` takes a
`https://kafka.apache.org/25/javadoc/org/apache/kafka/clients/consumer/KafkaConsumer.html[KafkaConsumer]`
and gives back a collection of `Book`.

> <2> `Consumer#consume` call the `KafkaConsumer#poll` and gets a `ConsumerRecords` filed with as much books as the
> `fetch.max.bytes` Consumer Config allows.

Reviewer 2: IMO remove the dialog about fetch.max.bytes; slightly advanced and also doesn't tell the full story about how many records will be in the return set.

```suggestion
<2> We call the `KafkaConsumer#poll` function on the `KafkaConsumer` which returns a `ConsumerRecords` containing Book instances.
```

> <3> `Consumer#consume` extracts the value from each records. The consumed books will filled `bookMap`, our book collection
> of type `mutable.Map[String, Book]`.

Reviewer 2: Considering the addition to the class level map isn't done here, IMO it's slightly confusing to the reader.  I would remove that bit:

```suggestion
<3> For each returned value in the `ConsumerRecords` instance, the record value is extracted and passed back in a `Vector`. 
```

> We also have two functions `getCount` and `getBooks` just to define two HTTP routes.
> 

Reviewer 2: ```suggestion
Our Consumer application functions as an HTTP service defining two routes, `/count` and `/books`.  These routes are mapped to the functions `getCount` and `getBooks` in our code.

```

> 🤔 🤔 Now what do we need to call our new `Consumer#consume` function?
> 
> - `reflectionDeserializer4S[T]` is introduced to lighten the tutorial and quickly get a `Deserializer[Book]`
> (See dev part n°3).
> 
> - We carefully configure, the deserializer with a map containing the schema-registry url.
> 
> - Then we instantiate the `KafkaConsumer` by passing a Java Properties from the _consumer.client-config_ block of
> `application.conf`.
> 
> - We subscribe to the book's topics by calling `KafkaConsumer#suscribe` with a topic name collection.
> 
> That's it! we are ready to poll records in a infinite while loop to keep our Map of book updated.

Reviewer 2: ```suggestion
Similar to the Producer code, we utilize our `Consumer#consume` function by constructing an instance of the `KafkaConsumer`.  

- The `KafkaConsumer` is constructed with the `reflectionDeserializer4S[T]` which is configured to connect to the Schema Registry.

- The `KafkaConsumer` reads it's configuration from the _consumer.client-config_ block of
`application.conf`.

- Finally, we subscribe to the `BOOK` topic by calling `KafkaConsumer#suscribe`.

That's it! we are ready to poll records in a loop, updating our map of books.  The map of books is used to respond to queries on the HTTP routes.
```
